symbolic-debuginfo/src/wasm.rs

```diff
@@ -0,0 +1,341 @@
+    pub fn code_id(&self) -> Option<CodeId> {
+        // see `debug_id`
+        for (_, section) in self.wasm_module.customs.iter() {
+            if section.name() == "build_id" {
+                return Some(CodeId::from_binary(&section.data(&Default::default())));
+            }
+        }
+        None
+    }
+
+    /// The debug information identifier of a WASM file.
+    ///
+    /// Wasm does not yet provide debug IDs.
+    pub fn debug_id(&self) -> DebugId {
+        for (_, section) in self.wasm_module.customs.iter() {
+            // this section is not defined yet
+            // see https://github.com/WebAssembly/tool-conventions/issues/133
+            if section.name() == "build_id" {
+                return DebugId::from_guid_age(
+                    section
+                        .data(&Default::default())
+                        .get(..16)
+                        .unwrap_or(&[][..]),
+                    0,
+                )
+                .ok()
+                .unwrap_or_default();
+            }
+        }
+        DebugId::nil()
+    }
+
+    /// The CPU architecture of this object.
+    pub fn arch(&self) -> Arch {
+        Arch::Wasm
+    }
+
+    /// The kind of this object.
+    pub fn kind(&self) -> ObjectKind {
+        if self.wasm_module.funcs.iter().next().is_some() {
+            ObjectKind::Library
+        } else {
+            ObjectKind::Debug
+        }
+    }
+
+    /// The address at which the image prefers to be loaded into memory.
+    ///
+    /// This is always 0 as this does not really apply to WASM.
+    pub fn load_address(&self) -> u64 {
+        0
+    }
+
+    /// Determines whether this object exposes a public symbol table.
+    pub fn has_symbols(&self) -> bool {
+        true
+    }
+
+    /// Returns an iterator over symbols in the public symbol table.
+    pub fn symbols(&self) -> WasmSymbolIterator<'data, '_> {
+        let iterator = Box::new(self.wasm_module.funcs.iter()) as Box<dyn Iterator<Item = _>>;
+        WasmSymbolIterator {
+            funcs: iterator.peekable(),
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Returns an ordered map of symbols in the symbol table.
+    pub fn symbol_map(&self) -> SymbolMap<'data> {
+        self.symbols().collect()
+    }
+
+    /// Determines whether this object contains debug information.
+    pub fn has_debug_info(&self) -> bool {
+        for (_, section) in self.wasm_module.customs.iter() {
+            if section.name() == ".debug_info" {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Constructs a debugging session.
+    pub fn debug_session(&self) -> Result<DwarfDebugSession<'data>, DwarfError> {
+        let symbols = self.symbol_map();
+        // WASM is offset by the negative offset to the code section instead of the load address
+        DwarfDebugSession::parse(self, symbols, -(self.code_offset() as i64), self.kind())
```

Comment:
the comment makes sense, but considering you are using `saturating_sub` in the other place, I wonder if an example can make all this clearer?

Reply:
I'm changing the saturating to wrapping which makes more sense to the rest of the system.